### PR TITLE
Implement generators

### PR DESCRIPTION
**Pull request by Mdrbhatti**

### CODE REVIEW

File: src/genobject.ts [hunk 1]
```diff
@@ -19,23 +19,24 @@ class Py_GeneratorObject extends Py_FuncObject implements Iterator, Iterable {
...
+        this.frame.genFrame = true;
         return None;
     }
+
+    public $__next__ = new Py_AsyncNativeFuncObject((t: Thread, f: IPy_FrameObj, args: IPy_Object[], kwargs: Py_Dict, cb: (rv: IPy_Object) => void) => {
+        // Callback is saved and invoked when generator frame encounters YIELD_VALUE
+        this.frame.cb = cb;
+        this.thread.framePush(this.frame);
+        // Mark as the generator frame
+        this.frame.genFrame = true;
```
Comment (perimosocordiae) [inline src/genobject.ts]: Shouldn't this be `true` already?

File: src/optable.ts [hunk 2]
```diff
@@ -225,7 +225,23 @@ optable[opcodes.PRINT_NEWLINE] = function(f: Py_FrameObject) {
...
+    // So we set up an EXCEPT_HANDLER, AFTER a generator has been exhausted, and then raise StopIteration which is immediately caught
+    if (f.genFrame) {
+        f.genFrame = false;
+
+        // If generator was called inside a loop,  
+        if (t.loop_exc_block.length > 0) {
+            (<Py_FrameObject>f.back).blockStack.push(t.loop_exc_block.pop());
+        }
+
+        var x = (<any> builtins)['$StopIteration'];
```
Comment (perimosocordiae) [inline src/optable.ts]: `x` is unused.

**Comment by perimosocordiae** [comment]
Other than my small comments, this looks good to me.

File: src/frameobject.ts [hunk 3]
```diff
@@ -133,6 +135,11 @@ class Py_FrameObject implements IPy_FrameObj {
                 console.log(`${t.stackDepth()}: ${opcodes[op]}`);
             }
             func(this, t);
+            // Pop the generator frame (whenever a YIELD_VALUE occurs)
+            if(this.returnToThread && this.genFrame){
+                t.framePop();
```
Comment (jvilk) [inline src/frameobject.ts]: Shouldn't `YIELD_VALUE` simply call `asyncReturn` to pop off its frame and give the value to the calling frame?

Comment (jvilk) [inline src/frameobject.ts]: Meaning: I don't think the `genFrame` boolean or this special case is even needed.

File: src/optable.ts [hunk 4]
```diff
@@ -984,19 +1004,35 @@ optable[opcodes.GET_ITER] = function(f: Py_FrameObject, t: Thread) {
...
+        } else {
+            f.pop();
+            f.lastInst += delta;
+        }
     }
 }
 
+optable[opcodes.YIELD_VALUE] = function(f: Py_FrameObject, t: Thread) {
+    var res = f.peek();
+    // Push on calling frame's stack
```
Comment (jvilk) [inline src/optable.ts]: I think you should just call `asyncReturn` here. `YIELD_VALUE` is essentially identical to `RETURN_VALUE`, except it leaves the generator in a usable state.

File: src/optable.ts [hunk 5]
```diff
@@ -909,16 +923,22 @@ optable[opcodes.STORE_MAP] = function(f: Py_FrameObject) {
...
+function setup_block(f: Py_FrameObject, t: Thread, op: number) {
     var delta = f.readArg();
     // push a block to the block stack
     var stackSize = f.stack.length;
     var loopPos = f.lastInst;
     f.blockStack.push([stackSize, loopPos, loopPos+delta, op]);
+
+    // For handling StopIteration exceptions inside of for loops, generated by exhausted generators
+    if (op === opcodes.SETUP_LOOP) {
+        t.loop_exc_block.push([stackSize, loopPos, loopPos + delta, opcodes.SETUP_EXCEPT]);
```
Comment (jvilk) [inline src/optable.ts]: Is this ever popped off?

Comment (jvilk) [inline src/optable.ts]: OK, I see, it is during a return value. Looks like these are transferred to the block stack. How is this different from the block stack?

File: src/optable.ts [hunk 6]
```diff
@@ -225,7 +225,21 @@ optable[opcodes.PRINT_NEWLINE] = function(f: Py_FrameObject) {
...
-    t.asyncReturn(f.pop());
+    // If a generator frame is exhausted, we raise StopIteration exception
+    // StopIteration exceptions generated by generators inside a for loop are ALWAYS caught
+    // So we set up an EXCEPT_HANDLER, AFTER a generator has been exhausted, and then raise StopIteration which is immediately caught
+    if (f.genFrame) {
+        f.genFrame = false;
+
+        // If generator was called inside a loop,  
+        if (t.loop_exc_block.length > 0) {
+            (<Py_FrameObject>f.back).blockStack.push(t.loop_exc_block.pop());
```
Comment (jvilk) [inline src/optable.ts]: Is it OK that you're reversing the order of these?

(e.g. if `loop_exc_block` contains values [1,2,3], you're pushing it onto `blockStack` as [3,2,1])

**Comment by jvilk** [comment]
I left some comments.

**Comment by jvilk** [comment]
LGTM. Time to merge!

**Pull request merged**